data/src/test/java/com/linkedin/data/schema/validation/TestValidation.java

```diff
@@ -2081,4 +2098,161 @@ public void testNonRootStartDataElement() throws IOException
+            "    { \"name\" : \"unionField\", \"type\" : [ \"int\", \"string\", \"Foo\" ], \"optional\" : true },\n" +
+            "    { \"name\" : \"recordField\", \"type\" : \"Foo\", \"optional\" : true }\n" +
+            "  ]\n" +
+            "}\n";
+
+    DataSchema schema = dataSchemaFromString(schemaText);
+
+    String dataString =
+        "{\n" +
+        "  \"primitive\" : 1,\n" +
+        "  \"unrecognizedPrimitive\": -1,\n" +
+        "  \"arrayField\" : [ { \"primitive\": 2, \"unrecognizedInArray\": -2 } ],\n" +
+        "  \"mapField\" : { \"key\": { \"primitive\": 3, \"unrecognizedInMap\": -3 } },\n" +
+        "  \"unionField\" : { \"Foo\": { \"primitive\": 4, \"unrecognizedInMap\": -4 } },\n" +
+        "  \"recordField\" : {\n" +
+        "    \"primitive\" : 5,\n" +
+        "    \"unrecognizedPrimitive\": -5\n" +
+        "  },\n" +
+        "  \"unrecognizedMap\": { \"key\": -100},\n" +
+        "  \"unrecognizedArray\": [ -101 ]\n" +
+        "}";
+
+    String trimmedDataString =
+        "{\n" +
+        "  \"primitive\" : 1,\n" +
+        "  \"arrayField\" : [ { \"primitive\": 2 } ],\n" +
+        "  \"mapField\" : { \"key\": { \"primitive\": 3 } },\n" +
+        "  \"unionField\" : { \"Foo\": { \"primitive\": 4 } },\n" +
+        "  \"recordField\" : {\n" +
+        "    \"primitive\" : 5\n" +
+        "  }\n" +
+        "}";
+
+    // mutable
+    DataMap toValidate = dataMapFromString(dataString);
+    ValidationResult result = validate(toValidate, schema, options);
+    assertTrue(result.isValid(), MessageUtil.messagesToString(result.getMessages()));
+    assertEquals(result.getFixed(), dataMapFromString(trimmedDataString));
+    assertSame(toValidate, result.getFixed());
+
+    // read-only
+    DataMap readOnlyToValidate = dataMapFromString(dataString);
+    readOnlyToValidate.makeReadOnly();
+    ValidationResult readOnlyResult = validate(readOnlyToValidate, schema, options);
+    assertTrue(readOnlyResult.hasFixupReadOnlyError());
+    String message = readOnlyResult.getMessages().toString();
+    assertTrue(readOnlyResult.getMessages().size() == 7);
+    String[] expectedStrings = new String[] {
+        "/unrecognizedMap",
+        "/unrecognizedArray",
+        "/recordField/unrecognizedPrimitive",
+        "/unionField/Foo/unrecognizedInMap",
+        "/unrecognizedPrimitive",
+        "/arrayField/0/unrecognizedInArray",
+        "/mapField/key/unrecognizedInMap"
+    };
+    for (String expected : expectedStrings)
+    {
+      assertTrue(message.contains(expected), message + " does not contain " + expected);
+    }
+  }
+
+  @Test
+  public void testUnrecognizedFieldsWithAvroUnionDefault() throws IOException
+  {
+    ValidationOptions options = new ValidationOptions(
+        RequiredMode.CAN_BE_ABSENT_IF_HAS_DEFAULT,
+        CoercionMode.NORMAL,
+        UnrecognizedFieldMode.TRIM);
```

Comment:
Can you write a test with disallow as well? It should be similar except the result would not be valid